doc/man7/provider-keymgmt.pod

```diff
@@ -219,13 +214,43 @@ returns NULL, the caller is free to assume that there's an algorithm
 from the same provider, of the same name as the one used to fetch the
 keymgmt and try to use that.
 
+OP_keymgmt_has() should check whether the given I<keydata> the subsets
+of data indicated by the I<selector>.  A combination of several
+selector bits must consider all those subsets, not just one.  An
+implementation is, however, free to consider an empty subset of data
+to still be a valid subset.
+
+OP_keymgmt_validate() should check if the I<keydata> contains valid
+data subsets indicated by I<selection>.  Some combined selections of
+data subsets may cause validation of the combined data.
+For example, the combination of B<OSSL_KEYMGMT_SELECT_PRIVATE_KEY> and
+B<OSSL_KEYMGMT_SELECT_PUBLIC_KEY> (or B<OSSL_KEYMGMT_SELECT_KEYPAIR>
+for short) is expected to check that the pairwise consistency of
+I<keydata> is valid.
+
+=head2 Key Object Import and Export Functions
+
+OP_keymgmt_import() should import data indicated by I<selection> into
+I<keydata> with values taken from the B<OSSL_PARAM> array I<params>.
+
+OP_keymgmt_export() should extract values indicated by I<selection>
+from I<keydata>, create an B<OSSL_PARAM> array with them and call
+I<param_cb> with that array as well as the given I<cbarg>.
+
+OP_keymgmt_import_types() should return a constant array of descriptor
+B<OSSL_PARAM> for data indicated by I<selection>, for parameters that
+OP_keymgmt_import() can handle.
+
+OP_keymgmt_export_types() should return a constant array of descriptor
+B<OSSL_PARAM> for data indicated by I<selection>, that the
+OP_keymgmt_export() callback can expect to receive.
+
```

Comment:
So looking at current `dh_kmgmt.c`, the implementation does not match the documentation.

Theoretically `libcrypto` can ask the provider to import only the private key in `keydata`, but that is not possible and no check is done to prevent it.

If the rule of thumb is that the `OSSL_PARAM` array resulting from `OP_keymgmt_export()` with `selection = FOO` must be at least capable of importing a valid `keydata` object when the same `OSSL_PARAM` array is passed to `OP_keymgmt_import()` with the same value for `selection`, this should be stated clearly.

We should also state what should the provider do if `libcrypto` sets `selection` to some combination that makes it impossible to do what stated above.
- should `OP_keymgmt_export()` ignore the restrictions set in `selection` and export whatever it wants?
- should `OP_keymgmt_export()` alternatively return an error as soon as an unsupported selection is requested?
- should `OP_keymgmt_import()` return an error as soon as an unsupported selection is requested?

The previous separation between `OP_keymgmt_importkey()`+`OP_keymgmt_exportkey()` and `OP_keymgmt_importdomparams()`+`OP_keymgmt_exportdomparams()` was IMHO less ambiguous and easier for provider implementers.

At this point I am not sure anymore when import and export can be triggered, but I see the potential risk for a key being generated in the provider, exported to `libcrypto` only partially because of `selection`, some `dirty_cnt` change is triggered in `libcrypto` which then triggers an import in the provider again resulting in loss of information or error.

Reply:
Uhmmm...  have you read what's written in [provider-keymgmt(7) / Key Object Import and Export Functions](https://github.com/openssl/openssl/blob/master/doc/man7/provider-keymgmt.pod#key-object-import-and-export-functions).  Doesn't "indicated by _selection_" tell you that _selection_ is not supposed to be ignored?

Of course, the provider code might ignore _selection_, but that seems rather dumb, doesn't it?  But most of all, how are we going to stop that?